File: utils/filter_converter.py
"""
Convert structured filter array to MongoDB query object (matches TS filterConverter.ts).
Used only if we ever query MongoDB directly; Gateway accepts filters as-is.
"""

from schemas import Filter


def _build_filter_condition(filter_type: str, value: object) -> dict:
    """Build a MongoDB filter condition based on filter type."""
    if filter_type == "ne":
        return {"$ne": value}
    if filter_type == "gt":
        return {"$gt": value}
    if filter_type == "lt":
        return {"$lt": value}
    if filter_type == "gte":
        return {"$gte": value}
    if filter_type == "lte":
        return {"$lte": value}
    if filter_type == "in":
        return {"$in": value if isinstance(value, list) else [value]}
    if filter_type == "nin":
        return {"$nin": value if isinstance(value, list) else [value]}
    if filter_type == "regex":
        return {"$regex": value, "$options": "i"}
    # eq or default
    return value  # type: ignore[return-value]


def convert_filters_to_mongo(filters: list[Filter]) -> dict:
    """Convert an array of filters to a MongoDB query object."""
    if not filters:
        return {}
    query: dict = {}
    for f in filters:
        field, ftype, value = f.field, f.type, f.value
        if field in query:
            existing = query[field]
            if isinstance(existing, dict) and not isinstance(existing, list):
                if "$and" not in query:
                    query["$and"] = [{field: existing}]
                query["$and"].append({field: _build_filter_condition(ftype, value)})
                del query[field]
        else:
            query[field] = (
                value if ftype == "eq" else _build_filter_condition(ftype, value)
            )
    return query

File: src/utils/filterConverter.ts
/**
 * Convert structured filter array to MongoDB query object
 */

import { Filter } from '../types.js';

/**
 * Convert an array of filters to a MongoDB query object
 */
export function convertFiltersToMongo(filters: Filter[]): any {
  if (!filters || filters.length === 0) {
    return {};
  }
  
  const query: any = {};
  
  for (const filter of filters) {
    const { field, type, value } = filter;
    
    // Handle multiple filters on the same field (AND logic)
    if (query[field]) {
      // If field already has a condition, we need to merge
      if (typeof query[field] === 'object' && !Array.isArray(query[field])) {
        // Merge with existing filter for same field using $and
        if (!query.$and) {
          query.$and = [{ [field]: query[field] }];
        }
        query.$and.push({ [field]: buildFilterCondition(type, value) });
        delete query[field];
      }
    } else {
      // Simple case: first filter on this field
      query[field] = type === 'eq' ? value : buildFilterCondition(type, value);
    }
  }
  
  return query;
}

/**
 * Build a MongoDB filter condition based on filter type
 */
function buildFilterCondition(type: string, value: any): any {
  switch (type) {
    case 'ne':
      return { $ne: value };
    case 'gt':
      return { $gt: value };
    case 'lt':
      return { $lt: value };
    case 'gte':
      return { $gte: value };
    case 'lte':
      return { $lte: value };
    case 'in':
      return { $in: Array.isArray(value) ? value : [value] };
    case 'nin':
      return { $nin: Array.isArray(value) ? value : [value] };
    case 'regex':
      return { $regex: value, $options: 'i' };
    case 'eq':
    default:
      return value;
  }
}
